compiler/rustc_hir_analysis/src/astconv/mod.rs

```diff
@@ -2234,15 +2230,157 @@ impl<'o, 'tcx> dyn AstConv<'tcx> + 'o {
+            let param_env = tcx.param_env(impl_);
+
+            let impl_ty = tcx.bound_type_of(impl_);
+            let impl_substs = self.fresh_item_substs(impl_, &infcx);
+            let impl_ty = impl_ty.subst(tcx, impl_substs);
+
+            let InferOk { value: impl_ty, obligations } =
+                infcx.at(&cause, param_env).normalize(impl_ty);
+
+            // Check that the Self-types can be related.
+            let Ok(InferOk { obligations: sub_obligations, value: () }) = infcx
+                .at(&ObligationCause::dummy(), param_env)
+                .define_opaque_types(false)
+                .sup(impl_ty, self_ty)
+            else {
+                continue;
+            };
+
+            // Check whether the impl imposes obligations we have to worry about.
+            let impl_bounds = tcx.predicates_of(impl_);
+            let impl_bounds = impl_bounds.instantiate(tcx, impl_substs);
+
+            let InferOk { value: impl_bounds, obligations: norm_obligations } =
+                infcx.at(&cause, param_env).normalize(impl_bounds);
+
+            let impl_obligations =
+                traits::predicates_for_generics(|_, _| cause.clone(), param_env, impl_bounds);
+
+            let candidate_obligations = impl_obligations
+                .chain(norm_obligations.into_iter())
+                .chain(obligations.iter().cloned());
+
+            let mut matches = true;
+
+            // Evaluate those obligations to see if they might possibly hold.
+            for o in candidate_obligations {
+                let o = infcx.resolve_vars_if_possible(o);
+                if !infcx.predicate_may_hold(&o) {
+                    matches = false;
+                    unsatisfied_predicates.push(o.predicate);
+                }
+            }
+
+            // Evaluate those obligations to see if they might possibly hold.
+            for o in sub_obligations {
+                let o = infcx.resolve_vars_if_possible(o);
+                if !infcx.predicate_may_hold(&o) {
+                    matches = false;
+                    unsatisfied_predicates.push(o.predicate);
+                }
+            }
+
+            if !matches {
+                continue;
+            }
+
+            self.check_assoc_ty(assoc_item, name, def_scope, block, span);
+
+            let ty::Adt(_, adt_substs) = self_ty.kind() else {
+                // FIXME(inherent_associated_types)
+                bug!("unimplemented: non-adt self of inherent assoc ty");
+            };
+
+            let item_substs =
+                self.create_substs_for_associated_item(span, assoc_item, segment, adt_substs);
+            // FIXME(inherent_associated_types): Check if the obligations arising from the
+            // where-clause & the bounds on the associated type and its parameters hold.
+            let ty = tcx.bound_type_of(assoc_item).subst(tcx, item_substs);
+            return Ok(Some((ty, assoc_item)));
```

Comment:
So, the problem here: what if multiple candidates may hold? This means we're just picking the first.

Reply:
I've added a FIXME. Could you give me an example where there are multiple candidates and we pick the first? I wasn't able to find such a case. Every time I tried to craft such an example, I faced error [E0592](https://doc.rust-lang.org/stable/error-index.html#E0592) (duplicate definitions). Does it require a nontrivial lifetime setup or specialized impls?

Reply:
```
#![feature(inherent_associated_types)]

struct Wrapper<T>(T);

impl Wrapper<i32> {
    type Foo = i32;
}

impl Wrapper<()> {
    type Foo = ();
}

fn main() {
    let x: Wrapper<_>::Foo = ();
}
```

?

Reply:
Thanks for the example! I didn't think about inferred types at all since they used to crash in my patch for the longest time. I now use the `<dyn AstConv>::infer_ctxt()` if available and I've rewritten the probing & confirmation code to report an ambiguity error in such cases. :)